tools/depcheck/src/main.rs

```diff
@@ -0,0 +1,272 @@
+                let crate_version = iter
+                    .next()
+                    .expect("There must be a version after the crate name!");
+                let crate_version = str::from_utf8(crate_version)
+                    .expect("Must be utf-8")
+                    .to_owned();
+                Some(DepSpec {
+                    crate_name,
+                    crate_version,
+                })
+            } else {
+                None
+            }
+        })
+        .collect();
+    spec.sort();
+    spec.dedup();
+
+    spec
+}
+
+/// Given a `cargo tree` invocation and the dependency sets to check, checks for any unlisted or duplicated deps
+///
+/// `dep_list_name_for_error` is the name of the const above to show in the error suggestion
+fn test_dep_list(
+    package: &str,
+    edge_kind: &str,
+    extra_args: &str,
+    sets: &[&BTreeSet<&str>],
+    dep_list_name_for_error: &str,
+) {
+    println!("Testing `cargo tree -p {package} -e {edge_kind} --no-default-features {extra_args}`");
+    let mut errors = Vec::new();
+    let dep_list = get_dep_list(package, edge_kind, extra_args);
+    for i in dep_list.windows(2) {
+        if i[0].crate_name == i[1].crate_name {
+            errors.push(format!(
+                "Found two versions for `{0}` ({1} & {2})",
+                i[0].crate_name, i[0].crate_version, i[1].crate_version
+            ));
+        }
+    }
+
+    'dep_loop: for i in dep_list {
+        let name = &i.crate_name;
+        for s in sets {
+            if s.contains(&**name) {
+                continue 'dep_loop;
+            }
+        }
+        errors.push(format!(
+            "Found non-allowlisted crate `{name}`, consider adding to \
+                             {dep_list_name_for_error} in tools/depcheck/src/allowlist.rs if intentional"
+        ));
+    }
+
+    if !errors.is_empty() {
+        eprintln!("Found invalid dependencies:");
+        for e in errors {
+            eprintln!("\t{e}");
+        }
+        process::exit(1);
+    }
+}
+
+fn main() {
+    let basic: BTreeSet<_> = BASIC_DEPS.iter().copied().collect();
+    let serde: BTreeSet<_> = EXTRA_SERDE_DEPS.iter().copied().collect();
+    let experimental: BTreeSet<_> = EXTRA_EXPERIMENTAL_DEPS.iter().copied().collect();
+    let lstm: BTreeSet<_> = EXTRA_LSTM_DEPS.iter().copied().collect();
+    let ryu: BTreeSet<_> = EXTRA_RYU_DEPS.iter().copied().collect();
+    let capi: BTreeSet<_> = EXTRA_CAPI_DEPS.iter().copied().collect();
+    let capi_build: BTreeSet<_> = EXTRA_CAPI_BUILD_DEPS.iter().copied().collect();
+    let capi_logging: BTreeSet<_> = EXTRA_CAPI_LOGGING_DEPS.iter().copied().collect();
+    let blob: BTreeSet<_> = EXTRA_BLOB_DEPS.iter().copied().collect();
+    let fs: BTreeSet<_> = EXTRA_FS_DEPS.iter().copied().collect();
+    let test: BTreeSet<_> = EXTRA_TEST_DEPS.iter().copied().collect();
+    test_dep_list("icu", "normal", "", &[&basic], "`BASIC_DEPS`");
```

Comment:
Thought: This could be structured in a more modular way. Right now it is very specific to certain feature sets.